smolyay/basis.py

```diff
@@ -0,0 +1,226 @@
+import abc
+import math
+
+import numpy
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid.
+
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @abc.abstractproperty
+    def points(self):
+        """The points of the basis function assigned to Smolyak indices"""
+        pass
+
+
+    @abc.abstractmethod
+    def __call__(self,x):
+        """Compute term of basis function
+        Returns the output of the basis function with input x
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        pass
+
+
+class ChebyshevFirstKind(BasisFunction):
+    r"""Basis function Chebyshev polynomials of the first kind
+    Chebyshev polynomials are a sequence of polynomials described by the
+    following recurrence relation:
+
+    ..math:
+        T_0(x) = 1
+        T_1(x) = x
+        T_{n+1}(x) = 2xT_n(x) - T_{n-1}(x)
+
+    ``n`` is the degree of the polynomial this object describes
+    ``points`` represents the extrema of the polynomial of this degree on
+    the domain [-1,1]
+
+    The extrema of the polynomial are calculated via the following
+    equation:
+
+    ..math:
+        x_{n,j}^* = -\cos\left((\frac{j-1}{n-1}\pi\right), j = 1,...,n
+        n = 2^{exactness-1}+1
+
+    Parameters
+    ----------
+    n : int
+        degree of the Chebyshev polynomial
+    """
+
+    def __init__(self,n):
+        super().__init__()
+        self._n = n
+
+    @property
+    def points(self):
+        """extrema of polynomial"""
+        if len(self._points) == 0:
+            self._compute_points()
+        return self._points
+
+    @property
+    def n(self):
+        """degree of polynomial"""
+        return self._n
+
+    @n.setter
+    def n(self,n):
+        if self._n != n:
+            self._n = n
+            self._points = []
+
+    def _compute_points(self):
```

Comment:
Is this function equivalent to:
https://numpy.org/doc/stable/reference/generated/numpy.polynomial.chebyshev.chebpts1.html#numpy.polynomial.chebyshev.chebpts1
?

Reply:
Not exactly. That function appears to print the new extrema added by every nth term. However, it doesn't include -1 or 1. 
I could modify _compute_points to use this. The check in make_nested_chebyshev_points line 217 wouldn't have to be done. However, since it doesn't include -1 and 1, I would need to make special considerations in _compute_points and make_nested_chebyshev_points to make sure they're added correctly.